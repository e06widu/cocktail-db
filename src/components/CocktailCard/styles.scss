@import "../../style/variables";
.card-container {
    border: 1px solid $white;
    padding: 20px;
    text-align: center;
    .thumb-image {
        width: 100%;
    }
    .c-name,
    .c-category {
        letter-spacing: 0.2px;
    }
    .c-name {
        margin-bottom: 10px;
    }
    .c-category {
        margin-top: 10px;
    }
    Button {
        margin-top: 10px;
    }
    .btn {
        .icon {
            margin-right: 5px;
        }
    }
    .already-in {
        margin-top: 20px;
    }
}

@media screen and (max-width: 280px) {
    h1 {
        margin-top: 0;
    }
    .c-name,
    .c-category {
        font-size: 14px;
    }
}

@media screen and (max-width: 480px) {}