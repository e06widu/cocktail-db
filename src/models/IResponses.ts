import { ICocktail } from "./ICocktail";

export type ISearchCocktailResponse = {
    drinks: ICocktail[];
};

export type IRandomCocktailResponse = {
    drinks: ICocktail[];
};
export type IMultipleRandomCocktailResponse = {
    drinks: ICocktail[];
};

