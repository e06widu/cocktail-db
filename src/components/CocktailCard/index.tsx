import React, { useEffect, useState } from 'react'
import { ICocktail } from '../../models/ICocktail'
import { ReactComponent as AddIcon } from '../../images/add-button.svg';
import { ReactComponent as RemoveIcon } from '../../images/trash.svg';

import './styles.scss'
import { useQuery, useQueryClient } from '@tanstack/react-query'
import Button from '../Button'
import Toast from '../Toast'

interface ICocktailCardProps {
    cocktail: ICocktail
    mode: 'home' | 'favorites' | 'search'
    handleRemoveFromCartToast?: (cocktail: ICocktail) => void
}

const CocktailCard: React.FC<ICocktailCardProps> = ({ cocktail, mode, handleRemoveFromCartToast }) => {
    const { data: favorites } = useQuery<ICocktail[]>(['cocktail', 'favorites']);
    const queryClient = useQueryClient();
    const [showToast, setShowToast] = useState<boolean>(false);

    const [isAlreadyInFavorite, setIsAlreadyInFavorite] = useState<boolean>(false)

    useEffect(() => {
        if (favorites)
            setIsAlreadyInFavorite(favorites.some(item => item.idDrink === cocktail.idDrink));
    }, [cocktail.idDrink, favorites]);

    useEffect(() => {
        if (showToast) {
            const timer = setTimeout(() => {
                setShowToast(false);
            }, 3000);

            return () => {
                clearTimeout(timer);
            };
        }
    }, [showToast]);

    const handleAddToCart = () => {
        queryClient.setQueryData<ICocktail[]>(['cocktail', 'favorites'], (oldData) =>
            [...(oldData || []), cocktail]);
        setShowToast(true);
    };

    const handleRemoveFromCart = () => {
        queryClient.setQueryData<ICocktail[]>(['cocktail', 'favorites'], (oldData) =>
            oldData?.filter((item) => item.idDrink !== cocktail.idDrink)
        );
        if (handleRemoveFromCartToast) {
            void handleRemoveFromCartToast(cocktail);
        }
    };

    return (
        <div className='card-container'>
            <div className='c-name'>{cocktail.strDrink}</div>
            {<img className='thumb-image' src={cocktail.strDrinkThumb}
                alt={cocktail.strDrink} loading="lazy" />}

            {mode === 'home' ? <div className='c-category'>{cocktail.strCategory}</div> : null}
            {mode === 'search' ?
                isAlreadyInFavorite ? <div className='already-in'>Already in favorites</div> :
                    <Button disabled={isAlreadyInFavorite} onClick={() => void handleAddToCart()}>
                        <div className='btn'>
                            <div className="icon"><AddIcon /></div>
                            <div className='text'>FAVORITE</div>
                        </div>
                    </Button>
                : null}
            {mode === 'favorites' ?
                <Button onClick={() => void handleRemoveFromCart()}>
                    <div className='btn'>
                        <div className="icon"><RemoveIcon /></div>
                        <div className='text'>REMOVE</div>
                    </div>
                </Button> : null}
            {showToast && <Toast message={`Cocktail ${cocktail.strDrink} added to favorite`} />}
        </div>
    )
}

export default CocktailCard
