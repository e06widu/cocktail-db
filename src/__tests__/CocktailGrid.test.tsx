import { render } from '@testing-library/react';
import { ICocktail } from '../models/ICocktail';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import CocktailCard from '../components/CocktailCard';


const queryClient = new QueryClient();

const cocktail: ICocktail = {
  idDrink: '1',
  strDrink: 'Mojito',
} as ICocktail;

describe('CocktailCard', () => {
    test('renders cocktail card', () => {
      const { container } = render(
        <QueryClientProvider client={queryClient}>
          <CocktailCard
            cocktail={cocktail}
            mode="home"
            handleRemoveFromCartToast={jest.fn()}
          />
        </QueryClientProvider>
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  
    
  });