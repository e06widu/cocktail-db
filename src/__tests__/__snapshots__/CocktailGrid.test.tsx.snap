// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CocktailCard renders cocktail card 1`] = `
<div
  class="card-container"
>
  <div
    class="c-name"
  >
    Mojito
  </div>
  <img
    alt="Mojito"
    class="thumb-image"
    loading="lazy"
  />
  <div
    class="c-category"
  />
</div>
`;
